libzmix/src/verifiable_encryption/cs_verifiable_encryption.rs

```diff
@@ -0,0 +1,781 @@
+        let mut n_mul_2 = n.try_clone()?; // n*2
+        n_mul_2 = n_mul_2.lshift1()?;
+        let g_prime = modulus.rand_range()?;
+        let g = g_prime.mod_exp(&n_mul_2, &modulus, Some(ctx))?;
+        Ok(Self {
+            g,
+            h: n.increment()?,    // h = n+1
+            n_by_4: n.rshift(2)?, // n/4
+            modulus,
+        })
+    }
+
+    /// self.g^exp % self.modulus
+    pub fn raise_to_g(
+        &self,
+        exp: &BigNumber,
+        ctx: Option<&mut BigNumberContext>,
+    ) -> UrsaCryptoResult<BigNumber> {
+        self.exponentiate(&self.g, exp, ctx)
+    }
+
+    /// self.h^exp % self.modulus
+    pub fn raise_to_h(
+        &self,
+        exp: &BigNumber,
+        ctx: Option<&mut BigNumberContext>,
+    ) -> UrsaCryptoResult<BigNumber> {
+        self.exponentiate(&self.h, exp, ctx)
+    }
+
+    /// exponentiate in this Paillier group meaning the result is taken modulo this group's order (modulus). base^exp % self.modulus
+    pub fn exponentiate(
+        &self,
+        base: &BigNumber,
+        exp: &BigNumber,
+        ctx: Option<&mut BigNumberContext>,
+    ) -> UrsaCryptoResult<BigNumber> {
+        base.mod_exp(exp, &self.modulus, ctx)
+    }
+
+    /// base^2 % self.modulus
+    pub fn sqr(
+        &self,
+        base: &BigNumber,
+        ctx: Option<&mut BigNumberContext>,
+    ) -> UrsaCryptoResult<BigNumber> {
+        match ctx {
+            Some(mut ctx) => base
+                .sqr(Some(&mut ctx))?
+                .modulus(&self.modulus, Some(&mut ctx)),
+            None => base.sqr(None)?.modulus(&self.modulus, None),
+        }
+    }
+
+    /// Return a random element modulo the group order, i.e. modulus
+    pub fn rand(&self) -> UrsaCryptoResult<BigNumber> {
+        self.modulus.rand_range()
+    }
+
+    /// Return a random element modulo sqrt(modulo)/4
+    pub fn rand_for_enc(&self) -> UrsaCryptoResult<BigNumber> {
+        self.n_by_4.rand_range()
+    }
+
+    /// if a > (n^2)/2 then n^2 - a else a
+    pub fn abs(
+        &self,
+        a: &BigNumber,
+        ctx: Option<&mut BigNumberContext>,
+    ) -> UrsaCryptoResult<BigNumber> {
+        let a = a.modulus(&self.modulus, ctx)?;
+        let modulus_by_2 = self.modulus.rshift(1)?;
+        if a > modulus_by_2 {
+            self.modulus.sub(&a)
+        } else {
+            Ok(a)
+        }
+    }
+}
+
+/// The entity that can decrypt the ciphertext is called Inspector. The term is borrowed from Idemix.
+impl CSInspector {
+    /// Create public and private key for encryption. Also initialize the Paillier group.
+    /// `num_messages` is the maximum number of messages that the public-private key will support.
+    /// Trying to encrypt more than `num_messages` messages will result in error. Encrypting less is fine.
+    /// Key Generation from section 3.2 of the Practical Verifiable Encryption ... paper
+    pub fn new(num_messages: usize) -> UrsaCryptoResult<Self> {
```

Comment:
Good point. I had forgotten about serialization altogether. Fixed.

Reply:
Thanks. I think I could now call `encrypt(...)` using a previously serialized key. However, `decrypt` leans on the CSInspector struct for the keys. It might require a new CSInspector constructor that takes a serialized private key as an argument.

Reply:
Sorry that i was rushing it, didn't even add a test. Fixed it now. Renamed `CSInspector` to `CSKeypair`, now both `encrypt` and `decrypt` don't need `self`. `cs_encryption_serialization_deserialization` is a test for serialization.

Reply:
I like that naming better too. Thanks.